src/mito2/src/read/range.rs

```diff
@@ -0,0 +1,259 @@
+                            index: file_index,
+                            row_group_index: row_group_index as i64,
+                        }],
+                        num_rows: DEFAULT_ROW_GROUP_SIZE,
+                    });
+                }
+            } else {
+                // If we don't known the number of row groups in advance, scan all row groups.
+                ranges.push(RangeMeta {
+                    time_range: file.time_range(),
+                    indices: smallvec![file_index],
+                    row_group_indices: smallvec![RowGroupIndex {
+                        index: file_index,
+                        row_group_index: ALL_ROW_GROUPS,
+                    }],
+                    // This may be 0.
+                    num_rows: file.meta_ref().num_rows as usize,
+                });
+            }
+        }
+    }
+
+    fn push_seq_mem_ranges(memtables: &[MemtableRef], ranges: &mut Vec<RangeMeta>) {
+        // For non append-only mode, each range only contains one memtable.
+        for (i, memtable) in memtables.iter().enumerate() {
+            let stats = memtable.stats();
+            let Some(time_range) = stats.time_range() else {
+                continue;
+            };
+            ranges.push(RangeMeta {
+                time_range,
+                indices: smallvec![i],
+                row_group_indices: smallvec![RowGroupIndex {
+                    index: i,
+                    row_group_index: ALL_ROW_GROUPS,
+                }],
+                num_rows: stats.num_rows(),
+            });
+        }
+    }
+
+    fn push_seq_file_ranges(
+        num_memtables: usize,
+        files: &[FileHandle],
+        ranges: &mut Vec<RangeMeta>,
+    ) {
+        // For non append-only mode, each range only contains one file.
+        for (i, file) in files.iter().enumerate() {
+            let file_index = num_memtables + i;
+            ranges.push(RangeMeta {
+                time_range: file.time_range(),
+                indices: smallvec![file_index],
+                row_group_indices: smallvec![RowGroupIndex {
+                    index: file_index,
+                    row_group_index: ALL_ROW_GROUPS,
+                }],
+                num_rows: file.meta_ref().num_rows as usize,
+            });
+        }
+    }
+}
+
+/// Groups ranges by time range.
+/// It assumes each input range only contains a file or a memtable.
+fn group_ranges_for_seq_scan(mut ranges: Vec<RangeMeta>) -> Vec<RangeMeta> {
+    if ranges.is_empty() {
+        return ranges;
+    }
+
+    // Sorts ranges by time range (start asc, end desc).
+    ranges.sort_unstable_by(|a, b| {
+        let l = a.time_range;
+        let r = b.time_range;
+        l.0.cmp(&r.0).then_with(|| r.1.cmp(&l.1))
+    });
+    let mut range_in_progress = None;
+    // Parts with exclusive time ranges.
+    let mut exclusive_ranges = Vec::with_capacity(ranges.len());
+    for range in ranges {
```

Comment:
Looks like we have similar code here https://github.com/GreptimeTeam/greptimedb/blob/02df58c92a89bbd7f10315326e1b98139fa153b8/src/mito2/src/read/seq_scan.rs#L625-L648

Reply:
I'll refactor SeqScan in the next PR. After that, we can remove this.